### PR TITLE
[distributed] Complete expired distributed checkpoint deprecations

### LINKED ISSUE #191398: [Distributed] Audit deprecations whose advertised removal versions have passed
### Problem

Current `main` identifies itself as PyTorch `2.14.0a0`, but several `torch.distributed` warnings still advertise removals or default changes after versions that have long passed:

- [fully_sharded_data_parallel.py lines 1224-1232](https://github.com/pytorch/pytorch/blob/5d242823c6de2bbc21d636b2995a38340bd2a16c/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1224-L1232): `optim_input` removal after 1.13.
- [fully_sharded_data_parallel.py lines 1245-1253](https://github.com/pytorch/pytorch/blob/5d242823c6de2bbc21d636b2995a38340bd2a16c/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1245-L1253): legacy optimizer state-dict API removal after 2.2.
- [state_dict.py lines 275-293](https://github.com/pytorch/pytorch/blob/5d242823c6de2bbc21d636b2995a38340bd2a16c/torch/distributed/checkpoint/state_dict.py#L275-L293): `submodules` behavior removal in 2.5.
- [state_dict.py lines 1362-1377](https://github.com/pytorch/pytorch/blob/5d242823c6de2bbc21d636b2995a38340bd2a16c/torch/distributed/checkpoint/state_dict.py#L1362-L1377): module-keyed model state dict removal in 2.5.
- [filesystem.py lines 655-666](https://github.com/pytorch/pytorch/blob/5d242823c6de2bbc21d636b2995a38340bd2a16c/torch/distributed/checkpoint/filesystem.py#L655-L666): `overwrite` default changing after 2.5.

Context checked at upstream commit `5d242823c6de2bbc21d636b2995a38340bd2a16c`.

### Reproduction

Exercise each deprecated path and inspect its `FutureWarning`; each warning gives users a lifecycle date that is already in the past while the behavior remains available.

### Expected behavior

Each deprecation should be resolved according to current compatibility policy: remove the obsolete path/default, schedule it with an accurate release target, or update the warning if the API is intentionally retained. The resulting behavior and migration guidance should be internally consistent.

### Impact

Stale deadlines make warnings unreliable for users and leave maintainers without a clear signal about whether compatibility branches are still required.

### Suggested regression test

For paths that are removed, update tests to assert the supported replacement behavior and the chosen failure mode. For paths retained on a new schedule, assert the revised warning category and message without embedding already-expired version promises.

cc @weifengpy @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @chauhang @mori360 @pragupta @ppwwyyxx @LucasLLC @pradeepfn @wanchaol @fduwjj @wz337 @wconstab @msaroufim @dcci @aditvenk @kapilsh

### PR DESCRIPTION
**Pull request by RohitRathore1**
## Summary

Complete distributed checkpoint deprecations whose advertised removal versions have long passed (current main is at 2.14):

- `submodules` parameter in `get_state_dict` (deprecated in 2.5): now raises `ValueError`. Removed `StateDictOptions.keep_submodule_prefixes` and `_StateDictInfo.submodule_prefixes`.
- Module-keyed model state dict `Dict[nn.Module, ...]` in `set_model_state_dict` (deprecated in 2.5): now raises `ValueError`. Only `Dict[str, ValueType]` accepted.
- `FileSystemWriter`/`FsspecWriter` `overwrite` default (warned after 2.5): changed from `True` to `False`.

FSDP deprecations are in a separate PR: https://github.com/pytorch/pytorch/pull/191712

Partially fixes https://github.com/pytorch/pytorch/issues/191398

## Test plan

- [x] `python test/distributed/checkpoint/test_state_dict.py -v -k deprecate` — OK
- [x] `python test/distributed/checkpoint/test_file_system_checkpoint_cpu.py -v` — OK
- [x] `lintrunner -a` — No lint issues

### CODE REVIEW

**Comment by RohitRathore1** [comment]
looking into these test failures

File: torch/distributed/checkpoint/state_dict.py [hunk 1]
```diff
@@ -284,12 +272,10 @@ def _verify_options(
     Verify the model and options passed by the user and generates _StateDictInfo.
     """
     if submodules:
-        warnings.warn(
-            "Getting submodules only model/optim state_dict is deprecated and "
-            "will be removed in 2.5. This feature can be achieved by manually "
-            "filtering out the state_dict returned from get_state_dict.",
-            FutureWarning,
-            stacklevel=2,
+        raise ValueError(
```
Comment (ivy-zhou) [inline torch/distributed/checkpoint/state_dict.py]: cc: @fegin is this OK? I would think that if we're going to deprecate this parameter we should just remove it from the calling API as well instead of failing at runtime.

Comment (fegin) [reply torch/distributed/checkpoint/state_dict.py]: @ivy-zhou You are right. We should just remove this argument.

**Review by fegin — COMMENTED** [review]
Please split the change to "overwrite" to another PR. I'm confident that the changes to state_dict are not going to be reverted. But I'm not so sure about the "overwrite" one. So let's split into two PRs in case we do need to revert.

**Comment by RohitRathore1** [comment]
> Please split the change to "overwrite" to another PR. I'm confident that the changes to state_dict are not going to be reverted. But I'm not so sure about the "overwrite" one. So let's split into two PRs in case we do need to revert.

here is the PR for this https://github.com/pytorch/pytorch/pull/192183

**Comment by RohitRathore1** [comment]
@ivy-zhou @fegin reverted the BC lint workarounds. submodules and keep_submodule_prefixes are now fully removed from the API as requested. Added suppress-bc-linter label since this is an intentional BC break as these parameters were deprecated since PyTorch 2.2 and we're now at 2.14.
